### PR TITLE
Fix memory leaks in the code

### PR DESCRIPTION
**Pull request by andysim**
## Description
#1532 identified problem with memory leaks.  Using cppcheck, I've found a number of possible candidates.  My idea is to slowly work through all of them and replace all alloc/dealloc pairs with an RAII solution.  Feel free to chime in if you disagree with the and design decisions.  I'll do one object at a time so allow easy reverting, so this will eventually need to be squashed.

## Todos
Notable points (developer or user-interest) that this PR has or will accomplish.
- [ ] Fix leaks identified in #1532 by avoiding explicit allocation and deallocation.

## Checklist
- [ ] Tests added for any new features
- [x] [All or relevant fraction of full tests run](http://psicode.org/psi4manual/master/build_planning.html#how-to-run-a-subset-of-tests)

## Status
- [x] Ready for review
- [x] Ready for merge

### CODE REVIEW

File: psi4/src/psi4/libmints/petitelist.h [hunk 1]
```diff
@@ -50,6 +50,9 @@ class IntegralFactory;
 class Matrix;
 class Dimension;
 
+using ShellMapType = std::vector<std::array<int, 8>>;
+static std::array<int, 8> pgZeros{0, 0, 0, 0, 0, 0, 0, 0};
```
Comment (susilehtola) [inline psi4/src/psi4/libmints/petitelist.h]: Should this be const?

**Comment by susilehtola** [comment]
Cool!

File: psi4/src/psi4/libmints/petitelist.h (hunk 1 above)
Comment (andysim) [reply psi4/src/psi4/libmints/petitelist.h]: it sure should - thanks!

**Comment by dgasmith** [comment]
This is heroes' work, Andy!

**Comment by jturney** [comment]
Nice!

**Comment by dgasmith** [comment]
@andysim I know this isn't done. But could we think about getting this PR in and keep working on this in subsequent PRs?

**Comment by andysim** [comment]
Thanks for the prompt - it had slipped off my radar, true to form.  I have a few hours to bash at this a bit more this morning so I'll see how far I can get before we pull it in.  I still have the cppcheck output handy, so I should be able to make some pretty decent progress.

**Comment by andysim** [comment]
Ok, so I've addressed everything that cppcheck found.  There are a number of false positives in that list, but I still fixed them regardless.  The full tests pass for me (except `nbody-convergence`, which needs a higher angmom build and `dcft8`, which segfaults).  Is anybody else seeing issues with `dcft8`?  I'm using the 2019 Intel compilers, so it's possible that there's an issue in my toolchain.  If others are seeing no problems in `dcft8` I'll try to debug it later today.

Unfortunately, cppcheck isn't too helpful with regards to the memory leaks that this PR was originally intended to address.  After it's pulled in I'll try and find some other valgrind-like tools to address those issues.

**Comment by JonathonMisiewicz** [comment]
`dcft8` passes on my master. Let me know what you find for memory leak tools, or if `dcft8` continues giving you trouble.

**Comment by andysim** [comment]
Thanks, @JonathonMisiewicz .  I fixed the bug and it appears to be a genuine problem - could you review my last commit please?  The DCFT tests all pass with it, so I assume it's fine.  Strangely enough, cppcheck didn't pick up on this one and we appear to have been getting lucky with our usual compilers accidentally zeroing out the memory.

All, this one's ready for review and merge.  I'm sure there's some overlap with the many LGTM commits, so I strongly suggest that this goes in after those PRs are all merged: that way I can take care of handling any merge conflicts as penance for my tardiness in finishing this cppcheck pass.  Also, I deliberately pushed each commit separately here in case we need to roll back any individual changes, so maybe squashing isn't the best way to go here.

**Comment by JonathonMisiewicz** [comment]
@andysim That isn't "luck". That's the troublesome code being only three days old. The problematic arrays had been constructed by `init_int_array`, but @robertodr [had me](https://github.com/psi4/psi4/pull/1624#discussion_r280180193) [switch that out](https://github.com/psi4/psi4/pull/1624/commits/a6dc9f0c602ad874ca96fc7db3afa97e90c3af12). Apparently I hadn't done it correctly.

Do you have a reference on when this kind of initialization is necessary? Memory management in C++ continues to elude me.

**Comment by andysim** [comment]
Ah, got it.  I don't have a good reference unfortunately, but a good rule of thumb is to assume that C++ is not initializing memory for you.  Assigning to an entire array can be quite expensive, so it's left to the programmer.  Moreover, you may be assigning values to it later, which would render any compiler-assigned initial value moot.  The old `init_int_array` and friends were inherited from Psi3: they take care of allocating *and* zeroing the memory for you (using `malloc` and then `memset`) so they made things a little more convenient for the programmer.

The modern C++ equivalent is to use `std::vector<int> myVec(nElements);`  That simply makes a vector of length `nElements`, which can be passed to any function expecting an `int *` by passing `myVec.data()` or `&myVec[0]`.  Note that even this way of constructing a vector doesn't initialize the data at all: to do that you need to either `std::fill(myVec.begin(), myVec.end(), 0);` or, better still, tell the compiler which value to use to construct the vector with `std::vector<int> myVec(nElements, 0);`  

You can see in my commit that I only zeroed the first (index 0) element.  The reason for that is because the loop over irreps starts from 1 in that code, so indices 1+ are correctly assigned.  Index 0 was not being touched, so it retained whatever value resulted from the call to `new` that created the array.  I hope this helps - please let me know if there's anything I can clarify at all.

**Pull request merged**